sstables/sstable_set.hh

```diff
@@ -24,84 +24,207 @@
+    std::optional<uint64_t> version_erased;
+};
+
+using versioned_sstable = seastar::lw_shared_ptr<versioned_sstable_data>;
+
+class incremental_selector_impl {
+public:
+    virtual ~incremental_selector_impl() {}
+    virtual std::tuple<dht::partition_range, std::vector<versioned_sstable>, dht::ring_position_view> select(const dht::ring_position_view&) = 0;
+};
+
+class sstable_set_impl {
+public:
+    virtual ~sstable_set_impl() {}
+    virtual std::vector<versioned_sstable> select(const dht::partition_range& range) const = 0;
+    virtual void insert(versioned_sstable sst) = 0;
+    virtual void erase(versioned_sstable sst) = 0;
+    virtual std::unique_ptr<incremental_selector_impl> make_incremental_selector() const = 0;
+};
 
 // Structure holds all sstables (a.k.a. fragments) that belong to same run identifier, which is an UUID.
 // SStables in that same run will not overlap with one another.
 class sstable_run {
-    sstable_list _all;
+    std::unordered_set<shared_sstable> _all;
 public:
+    sstable_run(std::unordered_set<shared_sstable>&&);
     void insert(shared_sstable sst);
     void erase(shared_sstable sst);
     // Data size of the whole run, meaning it's a sum of the data size of all its fragments.
     uint64_t data_size() const;
-    const sstable_list& all() const { return _all; }
+    const std::unordered_set<shared_sstable>& all() const { return _all; }
 };
 
-class sstable_set {
+class sstable_set_snapshot;
+class version_reference;
+
+class sstable_list {
+    lw_shared_ptr<const std::unordered_set<versioned_sstable>> _versioned_sstables;
+    lw_shared_ptr<std::unordered_set<shared_sstable>> _unversioned_sstables;
+    uint64_t _size;
+    uint64_t _version;
+    friend class sstable_set;
+public:
+    sstable_list();
+    sstable_list(lw_shared_ptr<const std::unordered_set<versioned_sstable>> versioned_sstables, uint64_t size, uint64_t version);
+    sstable_list(const sstable_list& sstl);
+    sstable_list(sstable_list&& sstl) noexcept;
+    sstable_list& operator=(const sstable_list& sstl);
+    sstable_list& operator=(sstable_list&& sstl) noexcept;
+
+    class const_iterator : public std::iterator<std::forward_iterator_tag, const shared_sstable> {
+        std::unordered_set<versioned_sstable>::const_iterator _it;
+        std::unordered_set<versioned_sstable>::const_iterator _versioned_end;
+        std::unordered_set<shared_sstable>::const_iterator _unversioned_it;
+        uint64_t _version;
+    public:
+        const_iterator(std::unordered_set<versioned_sstable>::const_iterator it,
+                       std::unordered_set<versioned_sstable>::const_iterator versioned_end,
+                       std::unordered_set<shared_sstable>::const_iterator unversioned_it,
+                       uint64_t version);
+        const_iterator& operator++();
+        const_iterator operator++(int);
+        const shared_sstable& operator*() const;
+        bool operator==(const_iterator it) const;
```

Comment:
Is there a reason for taking this iterator by value? It's quite heavy, a const reference should suffice

Reply:
that is my mistake, wil fix